File: salmon_king/data/species.py
"""Salmon species tables. Prices are Kodiak prelim ex-vessel by year, not live quotes."""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum


class SpeciesId(StrEnum):
    KING = "king"
    RED = "red"
    PINK = "pink"
    CHUM = "chum"
    SILVER = "silver"


@dataclass(frozen=True)
class Species:
    id: SpeciesId
    common: str
    local: str
    scientific: str
    avg_lb: float  # Kodiak commercial mean dressed-ish weight
    # Peak day-of-year and width (game-shaped curves, timed to real run windows)
    peak_doy: int
    width_days: float
    notes: str


SPECIES: dict[SpeciesId, Species] = {
    SpeciesId.KING: Species(
        id=SpeciesId.KING,
        common="Chinook",
        local="king",
        scientific="Oncorhynchus tshawytscha",
        avg_lb=5.3,  # incidental Kodiak fish; not SE troll kings
        peak_doy=165,  # mid-June
        width_days=16,
        notes="No directed commercial Chinook in the KMA. Incidental, small, scarce. Karluk/Ayakulik stocks of concern.",
    ),
    SpeciesId.RED: Species(
        id=SpeciesId.RED,
        common="sockeye",
        local="red",
        scientific="Oncorhynchus nerka",
        avg_lb=4.8,
        peak_doy=185,  # early July; late Karluk/Upper Station is a second pulse in-sim
        width_days=22,
        notes="Westside: Karluk-bound fish passing Central Section beaches. Alitak: Frazer / Upper Station / Akalura.",
    ),
    SpeciesId.PINK: Species(
        id=SpeciesId.PINK,
        common="pink",
        local="humpy",
        scientific="Oncorhynchus gorbuscha",
        avg_lb=3.2,
        peak_doy=210,  # late July
        width_days=20,
        notes="Two-year cycle; odd- and even-year lines are genetically distinct. Odd years flood; even years thin.",
    ),
    SpeciesId.CHUM: Species(
        id=SpeciesId.CHUM,
        common="chum",
        local="dog",
        scientific="Oncorhynchus keta",
        avg_lb=6.6,
        peak_doy=205,
        width_days=18,
        notes="July–August mixed-bag fish. Firm when fresh; go soft if you soak them.",
    ),
    SpeciesId.SILVER: Species(
        id=SpeciesId.SILVER,
        common="coho",
        local="silver",
        scientific="Oncorhynchus kisutch",
        avg_lb=6.5,
        peak_doy=245,  # early September
        width_days=18,
        notes="Mid-August through early October. Skeleton-crew fish after the pink push.",
    ),
}

# Kodiak preliminary average $/lb by calendar year (ADF&G season summaries).
# Used as June opening estimates. In-season the tender walks them around.
# 2026 is a generated year, not a live quote.
YEAR_PRICES: dict[int, dict[SpeciesId, float]] = {
    2023: {
        SpeciesId.KING: 0.57,
        SpeciesId.RED: 0.84,
        SpeciesId.SILVER: 0.28,
        SpeciesId.PINK: 0.26,
        SpeciesId.CHUM: 0.26,
    },
    2024: {
        SpeciesId.KING: 0.25,
        SpeciesId.RED: 1.16,
        SpeciesId.SILVER: 0.56,
        SpeciesId.PINK: 0.29,
        SpeciesId.CHUM: 0.25,
    },
    2025: {
        SpeciesId.KING: 0.10,
        SpeciesId.RED: 1.34,
        SpeciesId.SILVER: 0.70,
        SpeciesId.PINK: 0.30,
        SpeciesId.CHUM: 0.39,
    },
}

# Fallback when the player types a year we don't have a table for.
EVEN_YEAR_PRICES: dict[SpeciesId, float] = {
    SpeciesId.KING: 0.25,
    SpeciesId.RED: 1.05,
    SpeciesId.SILVER: 0.50,
    SpeciesId.PINK: 0.27,
    SpeciesId.CHUM: 0.28,
}
ODD_YEAR_PRICES: dict[SpeciesId, float] = {
    SpeciesId.KING: 0.34,
    SpeciesId.RED: 1.09,
    SpeciesId.SILVER: 0.49,
    SpeciesId.PINK: 0.28,
    SpeciesId.CHUM: 0.32,
}


def prices_for_year(year: int) -> dict[SpeciesId, float]:
    if year in YEAR_PRICES:
        return dict(YEAR_PRICES[year])
    return dict(ODD_YEAR_PRICES if year % 2 else EVEN_YEAR_PRICES)


SHORT = {
    SpeciesId.KING: "KNG",
    SpeciesId.RED: "RED",
    SpeciesId.PINK: "PNK",
    SpeciesId.CHUM: "CHM",
    SpeciesId.SILVER: "SLV",
}

File: src/data/species.rs
//! Salmon species tables. Prices are Kodiak prelim ex-vessel by year, not live quotes.

use crate::sim::models::Prices;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum SpeciesId {
    King = 0,
    Red = 1,
    Pink = 2,
    Chum = 3,
    Silver = 4,
}

impl SpeciesId {
    pub const ALL: [Self; 5] = [Self::King, Self::Red, Self::Pink, Self::Chum, Self::Silver];

    pub fn as_str(self) -> &'static str {
        match self {
            Self::King => "king",
            Self::Red => "red",
            Self::Pink => "pink",
            Self::Chum => "chum",
            Self::Silver => "silver",
        }
    }

    pub fn idx(self) -> usize {
        self as usize
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Species {
    pub id: SpeciesId,
    pub common: &'static str,
    pub local: &'static str,
    pub scientific: &'static str,
    pub avg_lb: f64,
    pub peak_doy: i32,
    pub width_days: f64,
    pub notes: &'static str,
}

pub const SPECIES: [Species; 5] = [
    Species {
        id: SpeciesId::King,
        common: "Chinook",
        local: "king",
        scientific: "Oncorhynchus tshawytscha",
        avg_lb: 5.3,
        peak_doy: 165,
        width_days: 16.0,
        notes: "No directed commercial Chinook in the KMA. Incidental, small, scarce. Karluk/Ayakulik stocks of concern.",
    },
    Species {
        id: SpeciesId::Red,
        common: "sockeye",
        local: "red",
        scientific: "Oncorhynchus nerka",
        avg_lb: 4.8,
        peak_doy: 185,
        width_days: 22.0,
        notes: "Westside: Karluk-bound fish passing Central Section beaches. Alitak: Frazer / Upper Station / Akalura.",
    },
    Species {
        id: SpeciesId::Pink,
        common: "pink",
        local: "humpy",
        scientific: "Oncorhynchus gorbuscha",
        avg_lb: 3.2,
        peak_doy: 210,
        width_days: 20.0,
        notes: "Two-year cycle; odd- and even-year lines are genetically distinct. Odd years flood; even years thin.",
    },
    Species {
        id: SpeciesId::Chum,
        common: "chum",
        local: "dog",
        scientific: "Oncorhynchus keta",
        avg_lb: 6.6,
        peak_doy: 205,
        width_days: 18.0,
        notes: "July–August mixed-bag fish. Firm when fresh; go soft if you soak them.",
    },
    Species {
        id: SpeciesId::Silver,
        common: "coho",
        local: "silver",
        scientific: "Oncorhynchus kisutch",
        avg_lb: 6.5,
        peak_doy: 245,
        width_days: 18.0,
        notes: "Mid-August through early October. Skeleton-crew fish after the pink push.",
    },
];

pub fn species(id: SpeciesId) -> &'static Species {
    &SPECIES[id.idx()]
}

/// Kodiak preliminary average $/lb by calendar year (ADF&G season summaries).
/// Used as June opening estimates. 2026 is a generated year, not a live quote.
pub fn prices_for_year(year: i32) -> Prices {
    match year {
        2023 => Prices {
            king: 0.57,
            red: 0.84,
            silver: 0.28,
            pink: 0.26,
            chum: 0.26,
        },
        2024 => Prices {
            king: 0.25,
            red: 1.16,
            silver: 0.56,
            pink: 0.29,
            chum: 0.25,
        },
        2025 => Prices {
            king: 0.10,
            red: 1.34,
            silver: 0.70,
            pink: 0.30,
            chum: 0.39,
        },
        y if y % 2 == 1 => Prices {
            king: 0.34,
            red: 1.09,
            silver: 0.49,
            pink: 0.28,
            chum: 0.32,
        },
        _ => Prices {
            king: 0.25,
            red: 1.05,
            silver: 0.50,
            pink: 0.27,
            chum: 0.28,
        },
    }
}

pub fn short_code(sp: SpeciesId) -> &'static str {
    match sp {
        SpeciesId::King => "KNG",
        SpeciesId::Red => "RED",
        SpeciesId::Pink => "PNK",
        SpeciesId::Chum => "CHM",
        SpeciesId::Silver => "SLV",
    }
}
